File: src/main/java/me/wawwior/utils/hashing/hashes/DefaultSaltedHash.java
package me.wawwior.utils.hashing.hashes;

import me.wawwior.utils.hashing.Hash;
import me.wawwior.utils.hashing.Salt;

import java.security.MessageDigest;
import java.util.Arrays;

public class DefaultSaltedHash implements Hash {

    private final MessageDigest digest;
    private final Salt salt;

    DefaultSaltedHash(MessageDigest digest, Salt salt) {
        this.digest = digest;
        this.salt = salt;
    }

    @Override
    public byte[] hash(byte[] bytes) {
        if (salt.getLength() > 0) {
            return hash(bytes, salt.getSalt());
        }
        return digest.digest(bytes);
    }

    private byte[] hash(byte[] bytes, byte[] salt) {
        byte[] salted = new byte[bytes.length + salt.length];
        System.arraycopy(bytes, 0, salted, 0, bytes.length);
        System.arraycopy(salt, 0, salted, bytes.length, salt.length);
        byte[] saltedHash = digest.digest(salted);
        byte[] result = new byte[saltedHash.length + salt.length];
        System.arraycopy(salt, 0, result, 0, salt.length);
        System.arraycopy(saltedHash, 0, result, salt.length, saltedHash.length);
        return result;
    }

    @Override
    public boolean compare(byte[] bytes, byte[] hash) {
        if (hash.length != hashLength())
            return false;
        if (salt.getLength() == 0) {
            return Arrays.equals(digest.digest(bytes), hash);
        }
        byte[] salt = new byte[this.salt.getLength()];
        System.arraycopy(hash, 0, salt, 0, salt.length);
        return Arrays.equals(hash(bytes, salt), hash);
    }

    @Override
    public int hashLength() {
        return digest.getDigestLength() + salt.getLength();
    }
}

File: src/main/kotlin/me/wawwior/utils/hashing/hashes/DefaultSaltedHash.kt
package me.wawwior.utils.hashing.hashes

import me.wawwior.utils.hashing.Hash
import me.wawwior.utils.hashing.Salt
import java.security.MessageDigest

class DefaultSaltedHash(private val messageDigest: MessageDigest, private val salt: Salt) : Hash {
    override fun hashLength() = messageDigest.digestLength + salt.getLength()

    override fun hash(bytes: ByteArray): ByteArray = if (salt.getLength() == 0) messageDigest.digest(bytes) else hash(bytes, salt.getSalt())

    private fun hash(bytes: ByteArray, salt: ByteArray): ByteArray {
        val saltedBytes = ByteArray(bytes.size + salt.size)
        System.arraycopy(bytes, 0, saltedBytes, 0, bytes.size)
        System.arraycopy(salt, 0, saltedBytes, bytes.size, salt.size)
        val saltedHash = messageDigest.digest(saltedBytes)
        val combinedHash = ByteArray(saltedHash.size + salt.size)
        System.arraycopy(salt, 0, combinedHash, 0, salt.size)
        System.arraycopy(saltedHash, 0, combinedHash, salt.size, saltedHash.size)
        return combinedHash
    }

    override fun compare(bytes: ByteArray, hash: ByteArray): Boolean {
        if (hash.size != hashLength())
            return false
        if (salt.getLength() == 0)
            return hash(bytes).contentEquals(hash)
        val salt = ByteArray(this.salt.getLength())
        System.arraycopy(hash, 0, salt, 0, salt.size)
        return hash(bytes, salt).contentEquals(hash)
    }
}